docs/content/config/best-practices/dkim_dmarc_spf.md

```diff
@@ -0,0 +1,299 @@
+
+1. Go inside the container with `docker exec -ti <CONTAINER NAME> bash`
+2. Run a command similar to `rspamadm dkim_keygen -s 'selector-name' -b 2048 -d example.com -k example.private > example.txt`, adjusted to your needs
+3. Make sure to then persists the files `example.private` and `example.txt` (created in step 2) in the container (for example with a Docker bind mount)
+4. Create a configuration for the DKIM signing module, i.e. a file called `dkim_signing.conf` that you mount to `/etc/rspamd/local.d/` or `/etc/rspamd/override.d/`. We provide example configurations down below. We recommend mounting this file into the container as well (as described [here](#manually)); do not use [`rspamd-modules.conf`](#with-the-help-of-a-custom-file) for this purpose.
+
+??? example "DKIM Signing Module Configuration Examples"
+
+    A simple configuration could look like this:
+
+    ```cf
+    # documentation: https://rspamd.com/doc/modules/dkim_signing.html
+
+    enabled = true;
+
+    sign_authenticated = true;
+    sign_local = true;
+
+    use_domain = "header";
+    use_redis = false; # don't change unless Redis also provides the DKIM keys
+    use_esld = true;
+    check_pubkey = true; # you wan't to use this in the beginning
+
+    domain {
+        example.com {
+            path = "/path/to/example.private";
+            selector = "selector-name";
+        }
+    }
+    ```
+
+    If you have multiple domains and you want to sign with the modern ED25519 elliptic curve but also with RSA (you will likely want to have RSA as a fallback!):
+
+    ```cf
+    # documentation: https://rspamd.com/doc/modules/dkim_signing.html
+
+    enabled = true;
+
+    sign_authenticated = true;
+    sign_local = true;
+
+    use_domain = "header";
+    use_redis = false; # don't change unless Redis also provides the DKIM keys
+    use_esld = true;
+    check_pubkey = true;
+
+    domain {
+        example.com {
+            selectors [
+                {
+                    path = "/path/to/com.example.rsa.private";
+                    selector = "dkim-rsa";
+                },
+                {
+                  path = /path/to/com.example.ed25519.private";
+                  selector = "dkim-ed25519";
+                }
+          ]
+        }
+        example.org {
+            selectors [
+                {
+                    path = "/path/to/org.example.rsa.private";
+                    selector = "dkim-rsa";
+                },
+                {
+                  path = "/path/to/org.example.ed25519.private";
+                  selector = "dkim-ed25519";
+                }
+            ]
+        }
+    }
+    ```
+
+!!! bug "File Permissions"
+
+    Make sure the user `_rspamd` is able to go into the directory where you persist the (private) key files, and ensure it can read them!
```

Comment:
For reference, the `open_dkim.bats` test would extract out the public key and concatentate it (_similar to what the DNS advice for web UI suggests further down_):

https://github.com/docker-mailserver/docker-mailserver/blob/adb38207ad35101bb01ce21621baa1c342fa205d/test/tests/serial/open_dkim.bats#L231-L243

[Smallstep CLI](https://smallstep.com/docs/step-cli/reference/crypto/keypair) can generate the keypairs (including Ed25519), while [another command](https://smallstep.com/docs/step-cli/reference/crypto/key/) from their CLI can inspect the private key or derive the public key from it (in [PEM format](https://stackoverflow.com/questions/57870914/how-to-create-a-single-line-x509-certificate-that-can-be-parsed-by-openssl-comma/58254058#58254058)). That would technically be capable of generating the main data for DKIM, the rest could be templated.

That said if OpenDKIM is going to be dropped by end of the year, `setup config dkim` could just replace opendkim for rspamd equivalent command :sweat_smile:

Reply:
Found [this awful looking monstrosity one-liner](https://serverfault.com/questions/466683/can-an-ssl-certificate-be-on-a-single-line-in-a-file-no-line-breaks/1055660#1055660) that reads in a public key in PEM format, excludes the header/footer lines and concatenates the base64 encoded content to a single line:

```bash
awk 'NR>2 { sub(/\r/, ""); printf "%s\\n",last} { last=$0 }' example.pub `
```

Probably don't want the `\n` being added there, so removing that and optionally using `fold` to revert back to the same PEM width:

```bash
awk 'NR>2 { sub(/\r/, ""); printf "%s",last} { last=$0 }' example.pub
  | fold --width 64
```

---

DKIM key + dns record generator helper:

```bash
#!/bin/bash

SELECTOR=${1:-mail}
KEYSIZE=${2:-2048}

# Create your own keypair instead of via `opendkim-genkey` or `rspamadm dkim_keygen`:
#step crypto keypair "${SELECTOR}.pub" "${SELECTOR}.private" --kty RSA --size "${KEYSIZE}" --no-password --insecure

# Derive the public key from the private key as a single line
PUBLIC_KEY=$(
  step crypto key public "${SELECTOR}.private" \
    | awk 'NR>2 { sub(/\r/, ""); printf "%s",last} { last=$0 }'
)

# Format into multiline 255 char segments for DNS record zone file:
PUBLIC_KEY_SPLIT=$(
  fold --width 255 <<< "p=${PUBLIC_KEY}" \
    | sed -E 's#(.*)#        "\1"#'
)

# Template of rspamadm output:
cat << EOF > "${SELECTOR}.dns.txt"
${SELECTOR}._domainkey IN TXT ( "v=DKIM1; k=rsa; "
${PUBLIC_KEY_SPLIT}
) ;
EOF
```

There you go, does not depend on opendkim or rspamd to generate the keypair and record, and can output the single line entry or multi-part DNS zone record.

`setup config dkim` could be retrofitted with that to generate a key in config volume that is stored in a common folder that both opendkim and rspamd could copy from at runtime and generate their configs for.